File: tests/e2e/wallet_test.go
package e2e_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"gitlab.com/2finance/2finance-network/blockchain/contract/contractV1/domain"
	"gitlab.com/2finance/2finance-network/blockchain/contract/walletV1"
	walletDomain "gitlab.com/2finance/2finance-network/blockchain/contract/walletV1/domain"
	walletModels "gitlab.com/2finance/2finance-network/blockchain/contract/walletV1/models"
	"gitlab.com/2finance/2finance-network/blockchain/log"
	"gitlab.com/2finance/2finance-network/blockchain/utils"
)

func TestWalletWorkflow(t *testing.T) {
	t.Helper()

	// ------------------
	//   LOCAL WALLET
	// ------------------
	wm := setupWalletManager(t)

	pub, priv := genKey(t, wm)
	importAndUnlockWallet(t, wm, pub, priv)

	c := setupClient(t, wm)

	// ------------------
	//   DEPLOY WALLET CONTRACT
	// ------------------
	deployedContract, err := c.DeployContract1(walletV1.WALLET_CONTRACT_V1)
	if err != nil {
		t.Fatalf("DeployContract: %v", err)
	}

	if len(deployedContract.Logs) == 0 {
		t.Fatalf("DeployContract returned no logs")
	}

	contractLog, err := utils.UnmarshalLog[log.Log](deployedContract.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (DeployContract.Logs[0]): %v", err)
	}

	assert.Equal(t, domain.DEPLOYED_CONTRACT_LOG, contractLog.LogType, "deploy log type mismatch")

	contractDomain, err := utils.UnmarshalEvent[domain.Contract](contractLog.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (DeployContract.Logs[0]): %v", err)
	}

	if contractDomain.Address == "" {
		t.Fatalf("contract address empty (event=%s)", string(contractLog.Event))
	}

	// ------------------
	//   ADD WALLET
	// ------------------
	wOut, err := c.AddWallet(contractDomain.Address, pub)
	if err != nil {
		t.Fatalf("AddWallet: %v", err)
	}

	if len(wOut.Logs) == 0 {
		t.Fatalf("AddWallet returned no logs")
	}

	walletLog, err := utils.UnmarshalLog[log.Log](wOut.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (AddWallet.Logs[0]): %v", err)
	}

	assert.Equal(t, walletDomain.WALLET_CREATED_LOG, walletLog.LogType, "add-wallet log type mismatch")

	wallet, err := utils.UnmarshalEvent[walletDomain.Wallet](walletLog.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (AddWallet.Logs[0]): %v", err)
	}

	assert.Equal(t, pub, wallet.PublicKey, "wallet public key mismatch")
	assert.Equal(t, contractDomain.Address, wallet.Address, "wallet address mismatch with contract address")

	// ------------------
	//   GET WALLET BY ADDRESS
	// ------------------
	wState, err := c.GetWalletByAddress(contractDomain.Address)
	if err != nil {
		t.Fatalf("GetWalletByAddress: %v", err)
	}

	if len(wState.Logs) != 0 {
		t.Fatalf("GetWalletByAddress returned logs")
	}

	if wState.States == nil || len(wState.States) == 0 {
		t.Fatalf("GetWalletByAddress returned no states")
	}

	walletState := walletModels.WalletStateModel{}

	err = utils.UnmarshalState[walletModels.WalletStateModel](wState.States[0].Object, &walletState)
	if err != nil {
		t.Fatalf("UnmarshalState (GetWalletByAddress.States[0]): %v", err)
	}

	assert.Equal(t, wallet.PublicKey, walletState.PublicKey, "wallet state public key mismatch between GetWalletByAddress and AddWallet event")
	assert.Equal(t, wallet.Address, walletState.Address, "wallet state address mismatch between GetWalletByAddress and AddWallet event")
	assert.NotEmpty(t, walletState.CreatedAt, "wallet state CreatedAt should not be empty")
	assert.NotEmpty(t, walletState.UpdatedAt, "wallet state UpdatedAt should not be empty")

	// ------------------
	//   GET WALLET BY PUBLIC KEY
	// ------------------
	wStateByPub, err := c.GetWalletByPublicKey(wallet.PublicKey)
	if err != nil {
		t.Fatalf("GetWalletByPublicKey: %v", err)
	}

	if len(wStateByPub.Logs) != 0 {
		t.Fatalf("GetWalletByPublicKey returned logs")
	}

	if wStateByPub.States == nil || len(wStateByPub.States) == 0 {
		t.Fatalf("GetWalletByPublicKey returned no states")
	}

	walletStateByPub := walletModels.WalletStateModel{}

	err = utils.UnmarshalState[walletModels.WalletStateModel](wStateByPub.States[0].Object, &walletStateByPub)
	if err != nil {
		t.Fatalf("UnmarshalState (GetWalletByPublicKey.States[0]): %v", err)
	}

	assert.Equal(t, walletState.PublicKey, walletStateByPub.PublicKey, "wallet state public key mismatch between GetWalletByAddress and GetWalletByPublicKey")
	assert.Equal(t, walletState.Address, walletStateByPub.Address, "wallet state address mismatch between GetWalletByAddress and GetWalletByPublicKey")
	assert.NotEmpty(t, walletStateByPub.CreatedAt, "wallet state CreatedAt should not be empty")
	assert.NotEmpty(t, walletStateByPub.UpdatedAt, "wallet state UpdatedAt should not be empty")
}

File: dart/blockchain/test/blockchain/contract/walletV2/wallet_test.dart
import 'dart:convert';

import 'package:two_finance_blockchain/blockchain/contract/contractV2/models/model.dart'
    as models;
import 'package:two_finance_blockchain/blockchain/contract/contractV2/domain/contract.dart'
    as domain;
import 'package:two_finance_blockchain/blockchain/contract/walletV2/constants.dart';
import 'package:two_finance_blockchain/blockchain/contract/walletV2/models/wallet.dart'
    as walletModels;
import 'package:two_finance_blockchain/blockchain/contract/walletV2/domain/wallet.dart'
    as walletDomain;
import 'package:two_finance_blockchain/blockchain/keys/keys.dart';
import 'package:two_finance_blockchain/blockchain/types/types.dart';
import 'package:two_finance_blockchain/blockchain/utils/json.dart';
import 'package:two_finance_blockchain/two_finance_blockchain.dart';
import 'package:two_finance_blockchain/blockchain/utils/marshal.dart';
import 'package:test/test.dart';
import 'package:web3dart/contracts.dart';
import '../../../helpers/helpers.dart';

void main() async {
  group('TwoFinanceBlockchain E2E - wallet', () {
    late TwoFinanceBlockchain c;

    setUpAll(() async {
      c = await setupClient();
    });

    tearDownAll(() async {
      await teardownClient(c);
    });

    test('generateKeyEd25519 + setPrivateKey deriva publicKeyHex', () async {
      final kp = await validKeyPair();

      // chaves geradas
      expect(kp.privateKey, isNotEmpty);
      expect(kp.publicKey, isNotEmpty);

      await c.setPrivateKey(kp.privateKey);

      // publicKeyHex derivada da private key
      expect(c.publicKeyHex, isNotNull);
      expect(c.publicKeyHex!, isNotEmpty);

      // validação oficial do seu pacote
      KeyManager.validateEDDSAPublicKeyHex(c.publicKeyHex!);

      // e deve bater com a pub gerada pelo helper (se seu validKeyPair for consistente)
      expect(c.publicKeyHex!, equals(kp.publicKey));
    });

    test(
      'E2E: createWallet (deploy + addWallet) + getWallet (roundtrip)',
      () async {
        // 1) signer
        final kp = await validKeyPair();
        await c.setPrivateKey(kp.privateKey);

        // 2) deploy wallet contract
        final deployedContract = await c.deployContract1(WALLET_CONTRACT_V2);

        final contractLogs = deployedContract.logs;
        expect(contractLogs, isNotNull);
        expect(contractLogs!, isNotEmpty);

        // pega o primeiro log
        final firstLog = contractLogs.first;

        // agora decodifica o event (base64 -> bytes) e faz unmarshalEvent
        final contractEvent = unmarshalEvent<domain.Contract>(
          firstLog.event,
          domain.Contract.fromJson,
        );

        expect(contractEvent.address, isNotEmpty);
        expect(contractEvent.contractVersion, equals(WALLET_CONTRACT_V2));
        print('Deployed contract at address: ${contractEvent.address}');
        print('Deployed public key: ${kp.publicKey}');

        final addWalletOutput = await c.addWallet(
          contractEvent.address,
          kp.publicKey,
        );
        expect(addWalletOutput, isNotNull);
        expect(addWalletOutput.logs, isNotNull);
        expect(addWalletOutput.logs!, isNotEmpty);

        final addWalletLog = addWalletOutput.logs!.first;
        expect(addWalletLog.logType, 'Wallet_Created');
        expect(addWalletLog.logIndex, 1);
        expect(addWalletLog.transactionHash, isNotEmpty);
        expect(addWalletLog.contractVersion, WALLET_CONTRACT_V2);
        expect(addWalletLog.contractAddress, contractEvent.address);
        expect(addWalletLog.event, isNotEmpty);

        final createdWallet = unmarshalEvent<walletDomain.Wallet>(
          addWalletLog.event,
          walletDomain.Wallet.fromJson,
        );

        expect(createdWallet.publicKey, equals(kp.publicKey));
        expect(createdWallet.address, equals(contractEvent.address));

        final getWalletOutput = await c.getWalletByAddress(
          contractEvent.address,
        );
        expect(getWalletOutput, isNotNull);
        expect(getWalletOutput.states, isNotNull);
        expect(getWalletOutput.states!, isNotEmpty);

        final getWalletStates = getWalletOutput.states;
        expect(getWalletStates, isNotNull);
        expect(getWalletStates!, isNotEmpty);

        final StateType state = getWalletStates!.first;

        final walletState = unmarshalState<walletModels.WalletState>(
          state.object,
          walletModels.WalletState.fromJson,
        );

        expect(walletState.publicKey, equals(kp.publicKey));
        expect(walletState.address, equals(contractEvent.address));
        expect(walletState.createdAt, isNotNull);
        expect(walletState.createdAt, isA<DateTime>());
        expect(walletState.updatedAt, isNotNull);
        expect(walletState.updatedAt, isA<DateTime>());
        expect(walletState.createdAt.isAfter(DateTime.utc(2000, 1, 1)), isTrue);
        expect(walletState.updatedAt.isAfter(DateTime.utc(2000, 1, 1)), isTrue);
        expect(
          walletState.updatedAt.isAfter(walletState.createdAt) ||
              walletState.updatedAt.isAtSameMomentAs(walletState.createdAt),
          isTrue,
        );
      },
    );
  });
}
